File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    const N = dungeons.length;
    const visited = new Array(N).fill(0);
    let answer = 0;

    function dfs(k, cnt){
        answer = Math.max(cnt, answer);

        for (let i = 0; i < N; i++) {
            if (k >= dungeons[i][0] && !visited[i]) {
                visited[i] = 1;
                dfs(k - dungeons[i][1], cnt + 1);
                visited[i] = 0;
            }
        }
    }

    dfs(k, 0);
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    
    def backtrack(cur_k, cnt):
        answer = cnt
        
        for i in range(len(dungeons)):
            k_req, cost = dungeons[i]
            if cur_k >= k_req and not visited[i]:
                visited[i] = True
                answer = max(answer, backtrack(cur_k - cost, cnt + 1))
                visited[i] = False
        
        return answer
    
    return backtrack(k, 0)